packages/loopring_v3/contracts/aux/agents/FastWithdrawalAgent.sol

```diff
@@ -12,152 +12,168 @@ import "../../lib/ReentrancyGuard.sol";
+    }
+
+    event Processed(
+        address exchange,
+        address from,
+        address to,
+        address token,
+        uint96  amount,
+        address provider,
+        address transactor,
+        Status  status
+    );
+
+    struct FastWithdrawalApproval
     {
         address exchange;
         address from;                   // The owner of the account
         address to;                     // The address that will receive the tokens withdrawn
         address token;
         uint96  amount;
-        address feeToken;
-        uint96  fee;
-        uint32  nonce;
-        uint32  validUntil;
-
-        bytes   signature;
+        uint32  storageID;
+        uint64  validUntil; // most significant 32 bits as block height, least significant 32 bits as block time
+        address provider;
+        bytes   signature; // provider's signature
     }
 
-    // EIP712
-    bytes32 constant public FASTWITHDRAWAL_TYPEHASH = keccak256(
-        "FastWithdrawal(address exchange,address from,address to,address token,uint96 amount,address feeToken,uint96 fee,uint32 nonce,uint32 validUntil)"
+    bytes32 constant public FASTWITHDRAWAL_APPROVAL_TYPEHASH = keccak256(
+        "FastWithdrawalApproval(address exchange,address from,address to,address token,uint96 amount,uint32 storageID,uint64 validUntil)"
     );
+
     bytes32 public DOMAIN_SEPARATOR;
 
     constructor()
     {
         DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
     }
 
-    // This method needs to be called by any liquidity provider
-    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
-        public
+    function execute(FastWithdrawalApproval memory fwa)
+        external
+        nonReentrant
+        payable
+    {
+        Status status = execute_(fwa);
+        emit Processed(
+            fwa.exchange,
+            fwa.from,
+            fwa.to,
+            fwa.token,
+            fwa.amount,
+            fwa.provider,
+            msg.sender,
+            status
+        );
+
+        msg.sender.sendETHAndVerify(address(this).balance, gasleft());
+    }
+
+    function execute(FastWithdrawalApproval[] memory fwas)
+        external
         nonReentrant
         payable
     {
-        // Do all fast withdrawals
-        for (uint i = 0; i < fastWithdrawals.length; i++) {
-            executeFastWithdrawal(fastWithdrawals[i]);
+        Status status;
+        for (uint i = 0; i < fwas.length; i++) {
+            status = execute_(fwas[i]);
+            emit Processed(
+                fwas[i].exchange,
+                fwas[i].from,
+                fwas[i].to,
+                fwas[i].token,
+                fwas[i].amount,
+                fwas[i].provider,
+                msg.sender,
+                status
+            );
         }
-        // Return any ETH left into this contract
-        // (can happen when more ETH is sent than needed for the fast withdrawals)
```

Comment:
ETH needs to be sent into the contract by the LP itself, and ETH can be stored in this contract, so it seems we currently can't let users do the fast withdrawal themselves when withdrawing ETH.

I still like the idea of msg.sender still always be the liquidity provider in the agent contract like it was before.

> We require the fast withdrawals to be executed by the liquidity provider (as msg.sender) so that the liquidity provider can impose its own rules on how its funds are spent.

This solves 2 issues:
- The ETH problem
- Much more flexible fund security for LPs. In this PR LPs are are forced to depend on a single signature of their hot wallet. Obviously that's not great when storing lots of funds there.

This still fully supports letting people submit their own fast withdrawals, it's just something the LP contract calling into the agent contract needs to support. No signatures would be checked directly in the agent contract (like before this PR).

Reply:
Brecht, could you show your implementation please?

Reply:
#1715